mlir/lib/Gradient/Transforms/vector_lowering.cpp

```diff
@@ -0,0 +1,229 @@
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Index/IR/IndexDialect.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/SCF/IR/SCF.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SymbolTable.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+#include "Gradient/IR/GradientOps.h"
+#include "Gradient/Transforms/Patterns.h"
+
+using namespace mlir;
+
+namespace catalyst {
+namespace gradient {
+/**
+ * A utility builder that aids in lowering dynamically-resizable parameter vectors.
+ *
+ * ParameterVectors are lowered to a size, a capacity, and a memref that stores the vector data.
+ * Each component is made mutable by being stored in a rank-0 memref, such that the vector data is
+ * lowered to (memref<memref<?x{element-type}>>, memref<index>, memref<index>).
+ */
+struct DynVectorBuilder {
+    Value dataField, sizeField, capacityField;
+    Type elementType;
+
+    static FailureOr<DynVectorBuilder> get(Location loc, TypeConverter *typeConverter,
+                                           TypedValue<ParameterVectorType> vector, OpBuilder &b)
+    {
+        SmallVector<Type> resultTypes;
+        if (failed(typeConverter->convertType(vector.getType(), resultTypes))) {
+            return failure();
+        }
+
+        auto unpacked = b.create<UnrealizedConversionCastOp>(loc, resultTypes, vector);
+        return DynVectorBuilder{.dataField = unpacked.getResult(0),
+                                .sizeField = unpacked.getResult(1),
+                                .capacityField = unpacked.getResult(2),
+                                .elementType = vector.getType().getElementType()};
+    }
+
+    FlatSymbolRefAttr getOrInsertPushFunction(Location loc, ModuleOp moduleOp, OpBuilder &b) const
+    {
+        MLIRContext *ctx = b.getContext();
+        std::string funcName = "__grad_vec_push";
+        llvm::raw_string_ostream nameStream{funcName};
+        nameStream << elementType;
+        if (moduleOp.lookupSymbol<func::FuncOp>(funcName))
+            return SymbolRefAttr::get(ctx, funcName);
+
+        OpBuilder::InsertionGuard guard(b);
+        b.setInsertionPointToStart(moduleOp.getBody());
+
+        auto pushFnType = FunctionType::get(
+            ctx, /*inputs=*/
+            {dataField.getType(), sizeField.getType(), capacityField.getType(), elementType},
+            /*outputs=*/{});
+        auto pushFn = b.create<func::FuncOp>(loc, funcName, pushFnType);
+        pushFn.setPrivate();
+
+        Block *entryBlock = pushFn.addEntryBlock();
+        b.setInsertionPointToStart(entryBlock);
+        BlockArgument elementsField = pushFn.getArgument(0);
+        BlockArgument sizeField = pushFn.getArgument(1);
+        BlockArgument capacityField = pushFn.getArgument(2);
+        BlockArgument value = pushFn.getArgument(3);
```

Comment:
Wouldn't it be better to get all arguments in one call from the performance POV?

Reply:
Sure, you're accessing arguments one by one in 4 function calls and I was wandering if this is possible/better to get all of them in one call. I'm not sure what the recommended practice is here or this is even possible though.

Reply:
`getArgument` here is a thin wrapper around an array access. The functions will probably be inlined, so I doubt the performance of these four function calls will be an issue.